src/Workspaces/SharedUtilitiesAndExtensions/Workspace/Core/Indentation/AbstractIndentationService.cs

```diff
@@ -0,0 +1,79 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Threading;
+using Microsoft.CodeAnalysis.Formatting;
+using Microsoft.CodeAnalysis.Formatting.Rules;
+using Microsoft.CodeAnalysis.Shared.Extensions;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis.Indentation
+{
+    internal abstract partial class AbstractIndentationService<TSyntaxRoot>
+        : AbstractIndentation<TSyntaxRoot>, IIndentationService
+        where TSyntaxRoot : SyntaxNode, ICompilationUnitSyntax
+    {
+        public IndentationResult GetIndentation(
+            Document document, int lineNumber,
+            FormattingOptions.IndentStyle indentStyle, CancellationToken cancellationToken)
+        {
+            if (indentStyle == FormattingOptions.IndentStyle.None)
+            {
+                // If there is no indent style, then do nothing.
+                return new IndentationResult(basePosition: 0, offset: 0);
+            }
+
+            var indenter = GetIndenter(document, lineNumber, (FormattingOptions2.IndentStyle)indentStyle, cancellationToken);
+
+            if (indentStyle == FormattingOptions.IndentStyle.Smart &&
+                indenter.TryGetSmartTokenIndentation(out var indentationResult))
+            {
+                return indentationResult;
+            }
+
+            // If the indenter can't produce a valid result, just default to 0 as our indentation.
+            return indenter.GetDesiredIndentation((FormattingOptions2.IndentStyle)indentStyle) ?? default;
+        }
+
+        private Indenter GetIndenter(Document document, int lineNumber, FormattingOptions2.IndentStyle indentStyle, CancellationToken cancellationToken)
+        {
+            var syntaxFormatting = this.SyntaxFormatting;
+
+#if CODE_STYLE
+            var tree = document.GetSyntaxTreeAsync(cancellationToken).WaitAndGetResult_CanCallOnBackground(cancellationToken);
```

Comment:
to me would read better if there was just a `GetTreeAndIndentation` and `GetIndentationRule` method that kind of hid the codestyle messiness, but that might just be me

Reply:
yeah.  i don't love this. but i don't feel like there are really good alternaives here :(